resqpy/grid/grid.py

```diff
@@ -0,0 +1,1235 @@
+        """Returns an integer array holding kji0 indices for the cells with given natural indices.
+
+        argument:
+           c0s: numpy integer array of shape (..., 3) being natural cell indices (for a flattened array)
+
+        returns:
+           numpy integer array of shape (..., 3) being the equivalent kji0 protocol cell indices
+        """
+
+        k0s, ji0s = divmod(c0s, self.nj * self.ni)
+        j0s, i0s = divmod(ji0s, self.ni)
+        return np.stack((k0s, j0s, i0s), axis = -1)
+
+    def resolve_geometry_child(self, tag, child_node = None):
+        """If xml child node is None, looks for tag amongst children of geometry root.
+
+        arguments:
+           tag (string): the tag of the geometry child node of interest
+           child_node (optional): the already resolved xml root of the child, or None
+
+        returns:
+           xml node of child of geometry node for this grid, which matches tag
+
+        note:
+           if child_node argument is not None, it is simply returned;
+           if child_node is None, the geometry node for this grid is scanned for a child with matching tag
+        """
+
+        if child_node is not None:
+            return child_node
+        return rqet.find_tag(self.geometry_root, tag)
+
+    def _set_k_raw_index_array(self):
+        """Sets the layering raw index array based on the k gap after boolean array."""
+        if self.k_gap_after_array is None:
+            self.k_raw_index_array = None
+            return
+        self.k_raw_index_array = np.empty((self.nk,), dtype = int)
+        gap_count = 0
+        for k in range(self.nk):
+            self.k_raw_index_array[k] = k + gap_count
+            if k < self.nk - 1 and self.k_gap_after_array[k]:
+                gap_count += 1
+        assert gap_count == self.k_gaps, 'inconsistency in k gap data'
+
+    def set_parent(self, parent_grid_uuid, self_is_refinement, parent_window):
+        """Set relationship with respect to a parent grid.
+
+        arguments:
+           parent_grid_uuid (uuid.UUID): the uuid of the parent grid
+           self_is_refinement (boolean): if True, this grid is a refinement of the subset of the parent grid;
+              if False, this grid is a coarsening
+           parent_window: (olio.fine_coarse.FineCoarse object): slice mapping information in K, j & I axes; note
+              that self_is_refinement determines which of the 2 grids is fine and which is coarse
+        """
+
+        if self.parent_grid_uuid is not None:
+            log.warning('overwriting parent grid information')
+        self.parent_grid_uuid = parent_grid_uuid
+        if parent_grid_uuid is None:
+            self.parent_window = None
+            self.is_refinement = None
+        else:
+            parent_window.assert_valid()
+            self.parent_window = parent_window
+            self.is_refinement = self_is_refinement
+
+    def actual_pillar_shape(self, patch_metadata = False, tolerance = 0.001):
```

Comment:
Could live in pillars.py but given that it is such a small method and would need a wrapper method anyway, perhaps not worth moving?